Trigger config refresh via ZookeeperTreeCachePropertySource

efenderbosch opened the pull request:
for issue #49

spencergibb commented on spring-cloud-zookeeper-config/src/main/java/org/springframework/cloud/zookeeper/config/ZookeeperConfigBootstrapConfiguration.java:
```diff
@@ -42,4 +48,35 @@ public ZookeeperPropertySourceLocator zookeeperPropertySourceLocator(
 	}
+
+	@Bean
+	public ApplicationListener<ContextRefreshedEvent> zookeeperTreeCachePropertySourceContextRefreshedApplicationListener() {
```
Regardless of the approach (still wrapping my head around it), this doesn't seem like the right way to set the application context. If it had to be set (again not sure), I'd set it when the property source is created.

efenderbosch replied:
When the property source is created, we only have access to the parent bootstrap context, not the child application context where all the real configuration lives. If there was a way to navigate down intto children contexts, then that would be sufficient.

efenderbosch commented on spring-cloud-zookeeper-config/src/main/java/org/springframework/cloud/zookeeper/config/ZookeeperTreeCachePropertySource.java:
```diff
@@ -110,4 +126,68 @@ public boolean isRunning() {
+	private static boolean shouldRefresh(String eventPath, String fullPath) {
+		// don't refresh if then lengths are equal because that means we got an
+		// event for the root config, not child data
+		return eventPath.startsWith(fullPath) && eventPath.length() == fullPath.length();
```
this should be length != length, missed that when I extracted the method

spencergibb commented:
@efenderbosch after some discussion, we'd like to stick with the approach taken elsewhere of responding to a global refresh event rather than what you have proposed. If you'd like to rework this PR, great, if not let's go ahead and close it.

efenderbosch commented:
Can I get a little insight in to that discussion? Or maybe I can layout why I think this approach might be better than a global refresh.

Let's say we have 4 microservices: ms1, ms2, m3 and ms4. We've implemented feature status checks (disabled, internal, beta, released) in our code to help with CI/CD. The features tree would live in a common area of the ZooKeeper tree so that it can be accessed by all services and changed in one place.

Release day is here and ms1 and ms2 will be deployed to release feature "two." ms3 and ms4 are unaffected by this feature.

Our ZooKeeper tree might look like:

/config/features/one: RELEASED
/config/features/two: DISABLED
/config/ms1/some_config/prop1: foo
/config/ms2/another_config/prop2: bar
/config/ms3/yet_another_config/prop3: baz
/config/ms4/one_more_config/prop4: qux

We release ms1 and ms2 and perform a smoke test. QA gives the all clear and we change /config/features/two from DISABLED to INTERNAL so they can begin another smoke test and then test feature "two." Instances of ms1 and ms2 correctly get their context refreshed, but instances of ms3 and ms4 are unnecessarily refreshed. With the approach I've implemented instances ms3 and ms4 would not have their contexts refreshed unnecessarily.

I'm also a bit concerned that various beans might not properly call close/shutdown in ``@PreDestroy`` methods leading to memory and/or resource leaks if all beans are refreshed globally.

Is the global refresh desirable because it is consistent with the Consul implementation? Or are there other concerns?

dsyer commented:
If you have concerns about `@PreDestroy` then please raise them (with a test case if you have one) in Spring Cloud Commons (AFAIK we test it there already, but there might be a bug I suppose).

Why do you care if a few beans get refreshed when nothing changed? Your scenario is certainly clear, but it's not really clear why it would be a problem.

I don't think anything in Consul is an issue, per se. It's more of a a Spring Cloud global design issue. We need all the config implementations to behave the same way ideally.

efenderbosch commented:
Fair enough. I guess was just concerned about the unnecessary bean refreshes. If that's not a real problem, then I'm good with the global refresh. I can make that change. Should I just close this PR and submit a new one?

dsyer commented:
I think an model based on `ApplicationEvent` is going to be the right one. If you can refactor this PR to do things that way, then we can start to think about merging. If you do a simple one-size-fits all model to start with (consistent with the other config providers) we can merge it immediately. If you want to somehow pick only the beans that need refreshing, we need to make changes in Spring Cloud Commons too probably (and maybe Spring Cloud Bus).

efenderbosch commented:
I can make it global like the Consul implementation for now. Might be a couple days before I can get back around to this.

efenderbosch commented:
Should it publish a ``EnvironmentChangeEvent`` from spring-cloud-commons? Then both spring-cloud-zookeeper and spring-cloud-consul could have a factory for those events and a listener to trigger the refresh.

spencergibb commented:
Superseded by 810fdad74995b971ca1242072ad08ff1a03c92c6
